### PR TITLE
DetailsList: Animate column updates

### PR DESCRIPTION
**Pull request by phkuo**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [x] Include a change request file using `$ npm run change`

#### Description of changes

Adds a simple animation for individual cells when that cell's content updates. Side effect: entire list also slightly animates when it first renders, but it actually looks kind of nice.

When new content comes in, the current cell's content immediately disappears, and the new contents immediately start sliding in from the left.

The old, super-complex PR is here:
https://github.com/OfficeDev/office-ui-fabric-react/pull/5384
That implementation also animated the sliding out of the old content before animating the sliding in of the new content. Holding on to that old content caused huge amounts of extra complexity, so I cut it.

#### Focus areas to test

perf

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/8936)

### CODE REVIEW

File: packages/office-ui-fabric-react/src/components/DetailsList/examples/DetailsList.Basic.Example.tsx [hunk 1]
```diff
@@ -43,6 +43,28 @@ export class DetailsListBasicExample extends React.Component<{}, IDetailsListBas
       });
     }
 
+    // DEMO CODE TODO: REMOVE BEFORE CHECKIN
```
Comment (phkuo) [inline packages/office-ui-fabric-react/src/components/DetailsList/examples/DetailsList.Basic.Example.tsx]: TODO TODO

**Comment by phkuo** [comment]
image
need to investigate

**Comment by phkuo** [comment]
> image
> need to investigate

@KevinTCoughlin I don't understand this one, it renders 100 items but is only expected to call back 10 times?

**Comment by KevinTCoughlin** [comment]
> @KevinTCoughlin I don't understand this one, it renders 100 items but is only expected to call back 10 times?

By default, the `List` component is virtualized with a page limit of size `10`.

https://github.com/OfficeDev/office-ui-fabric-react/blob/1e49aecddd978095b2bd60eb0307e683f7c31d49/packages/office-ui-fabric-react/src/components/List/List.tsx#L21
When I wrote this test it was to verify that although `100` items are supplied, the initial render cell call only happens `10` times.

It might be a flaky test, recently @jdhuntington had it fail on the `fabric-7` branch. @jdhuntington did you ever figure out what happened there or did it magically resolve?

**Comment by KevinTCoughlin** [comment]
@phkuo, a few questions:

1. Is this feature part of a design spec that `DetailsList` needs to match?
1. If so, can it be accomplished with the existing `styles` API? If not, what would we need to add to support that? It seems like we need to offer a way to trigger animations.

A more flexible / extensible approach is top of mind because folks may want to customize the animation and it would be easier for us to support long-term.

If we hard-code this prop and animation then it becomes part of our public contract and needs to be supported indefinitely.

cc: @aditima

**Comment by KevinTCoughlin** [comment]
I do like this level of customization as a feature. I wonder if we can make the API a bit more flexible to support other animations folks may want to provide. Perhaps treat the presence of an `IStyle` or similar as opt-in.

👍 to the perf callout. I can help take a look at that as part of review.

Thanks for sharing the comparison PR, https://github.com/OfficeDev/office-ui-fabric-react/pull/8936.

**Comment by dzearing** [comment]
I am totally hesitant to regress perf in any way. @KevinTCoughlin what do we want to do here, will this regress performance? We should avoid doing anything that goes in the wrong direction.

**Comment by KevinTCoughlin** [comment]
Two concerns that make me hesitant to move forward with this as-is:

1) There are incorrect borders on scroll of virtualized content (blocker):

Test URL: http://fabricweb.z5.web.core.windows.net/pr-deploy-site/refs/pull/8936/merge/fabric-website-resources/dist/index.html#/tests/detailslistbasicexample

phkuo_animate_cell

2) Intentional cell animations happening at rest confused me with `DetailsList.Basic.Example`

We use this example a lot to debug / dev on the `DetailsList` and `List` controls, so I'd avoid adding this behavior by default.

phkuo_animate_cell_at_rest

I haven't dug into the first issue... Nothing immediately stood out in the perf snapshots, but I need to look harder as I imagine there's impact on scroll. I'll follow-up.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx [hunk 2]
```diff
@@ -47,9 +47,19 @@ export class DetailsRowFields extends React.PureComponent<IDetailsRowFieldsProps
               ? onRender(item, itemIndex, column)
               : getCellText(item, column);
 
+          // generate a key that auto-dirties when content changes, to force the container to re-render, to trigger animation
+          let key: number | string;
```
Comment (ThomasMichon) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: I strongly recommend that you provide a way for the `key` to be provided to the `DetailsRow` directly via some pathway, so apps can tailor the value to suit their needs. Most of the time, the cell content is a JSX element, which might simply be a generic class component and not the value.

**Comment by phkuo** [comment]
I'll be passing this feature off to the Lists team. It sounds like the work left to be done is:
1) Have the key be passed in by the consumer, instead of decided on by DetailsList.
2) Determine how we want to handle the animation. Should the consumer have to pass it in, or do we put it in DetailsList and let the consumer opt in/out?
3) Since the animation is triggered by re-rendering, how do we prevent virtualization from triggering the animation? This would be easier to handle if the animation were within DetailsList instead of being passed in. (For instance, when it renders only put the animation class on if the key changes.)

**Pull request closed without merging**